working/0107 - implicit-constructors/feature-specification.md

```diff
@@ -0,0 +1,603 @@
+can contain constructors.*
+
+A compile-time error occurs if a static extension has an on-clause of the
+form `on C<T1 .. Tk>`, and the actual type parameters passed to `C` do not
+satisfy the bounds declared by `C`.  The static extension may declare one
+or more type parameters `X1 extends B1 .. Xs extends Bs`, and these type
+variables may occur in the types `T1 .. Tk`. During the bounds check, the
+bounds `B1 .. Bs` are assumed for the type variables `X1 .. Xs`.
+
+A compile-time error occurs if a static extension _D_ declares a
+constructor with the same name as a constructor in the on-class of _D_.
+*In other words, a constructor in a static extension can never have a name
+clash with a constructor declared by its on-class.*
+
+A compile-time error occurs if a static extension declares an
+`implicit` constructor whose formal parameter list accepts a number of
+positional paremeters which is different from one, or if it accepts any
+named parameters.
+
+Consider a static extension declaration _D_ named `E` which is declared in
+the current library or present in any exported namespace of an import
+*(that is, _D_ is declared in the current library or it is imported and
+not hidden, but it could be imported and have a name clash with some other
+imported declaration)*. A fresh name `FreshE` is created, and _D_ is
+entered into the library scope with the fresh name.
+
+*This means that a `static extension` declaration gets a fresh name in
+addition to the declared name, just like `extension` declarations.*
+
+*This makes it possible to resolve an implicit reference to a static
+extension, e.g., an invocation of a static member or constructor declared
+by the static extension, even in the case where there is a different
+declaration in scope with the same name. For example:*
+
+```dart
+static extension E on C { // `FreshE` is an extra name for `E`.
+  static void foo() {}
+}
+
+void f<E>() {
+  // Name clash: type parameter shadows static extension.
+  C.foo(); // Resolved as `FreshE.foo()`.
+}
+```
+
+#### Static extension scopes
+
+Static extensions introduce several scopes:
+
+The current scope for the on-clause of a static extension declaration _D_
+that does not declare any type parameters is the enclosing scope of _D_,
+that is the library scope of the current library.
+
+A static extension _D_ that declares type variables introduces a type
+parameter scope whose enclosing scope is the library scope. The current
+scope for the on-clause of _D_ is the type parameter scope.
+
+A static extension _D_ introduces a body scope, which is the current scope
+for the member declarations. The enclosing scope for the body scope is the
+type parameter scope, if any, and otherwise the library scope.
+
+Static members in a static extension are subject to the same static
+analysis as static members in other declarations. There is one extra rule:
+It is a compile-time error if a static member of a static extension has a
```

Comment:
As above, this makes it a breaking change to add a static member to a class.

Reply:
It is no more a breaking change.